clippy_lints/src/methods.rs

```diff
@@ -1562,6 +1583,21 @@ fn lint_map_unwrap_or(cx: &LateContext<'_, '_>, expr: &hir::Expr, map_args: &[hi
     }
 }
 
+/// lint use of `map().flatten()` for `Iterators`
+fn lint_map_flatten<'a, 'tcx>(
+    cx: &LateContext<'a, 'tcx>,
+    expr: &'tcx hir::Expr,
+    _map_args: &'tcx [hir::Expr],
+    _flatten_args: &'tcx [hir::Expr],
+) {
+    // lint if caller of `.map().flatten()` is an Iterator
+    if match_trait_method(cx, expr, &paths::ITERATOR) {
+        let msg = "called `map(..).flatten()` on an `Iterator`. \
+                   This is more succinctly expressed by calling `.flat_map(..)`";
+        span_lint(cx, MAP_FLATTEN, expr.span, msg);
```

Comment:
It would be nice if this generated suggestions like [here](https://github.com/rust-lang-nursery/rust-clippy/pull/3191/files#diff-a23361ccff4a035e853d0c993d8e1411R182)